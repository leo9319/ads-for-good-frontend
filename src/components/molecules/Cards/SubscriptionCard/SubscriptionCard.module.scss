@import '../../../../styles/colours';
@import '../../UserProfile/UserProfile.scss';

.cardContainer {
  width: 248px;
  height: 331px;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  background-color: $color-white;
  box-shadow: 0 1px 3px 0 $color-box-shadow;
  border: 1px solid $color-neutral-100;

  --card-border-width: 0;
  --card-padding: 0;
  --card-border-radius: var(--spacing-xl);
  --card-background-color: unset;

  &.enabledClick {
    cursor: pointer;
  }

  .imageBlock {
    width: 100% !important;
    height: 100% !important;
    object-fit: cover !important;
  }

  .profileBlock {
    display: grid;
    gap: var(--spacing-md);
    width: 100%;
    padding: var(--spacing-sm) var(--spacing-md);

    .head {
      display: grid;
      gap: var(--spacing-2xs);

      .childName {
        @include apply-text-styles(
          --typography-font-size-5,
          --typography-line-height-5
        );
      }

      .since {
        @include apply-text-styles(
          --typography-font-size-2,
          --typography-line-height-2,
          --typography-font-weight-regular,
          $color-neutral-600
        );
      }
    }

    .foot {
      display: grid;
      gap: var(--spacing-xs);

      .location {
        display: flex;
        gap: var(--spacing-xs);

        .content {
          @include apply-text-styles(
            --typography-font-size-2,
            --typography-line-height-2,
            --typography-font-weight-regular
          );
        }
      }

      .birthDate {
        display: flex;
        gap: var(--spacing-xs);

        .content {
          @include apply-text-styles(
            --typography-font-size-2,
            --typography-line-height-2,
            --typography-font-weight-regular
          );
        }
      }
    }
  }
}
