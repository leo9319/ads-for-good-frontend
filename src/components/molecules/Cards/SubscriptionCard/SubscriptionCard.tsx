import React, { ReactElement } from 'react';
import Icon from '@components/atoms/Icons';
import Image from '@components/atoms/Image';
import Box from '@radix-styles/atoms/Box';
import Card from '@radix-styles/atoms/Card';
import Heading from '@radix-styles/atoms/Heading';
import Text from '@radix-styles/atoms/Text';
import { classNames } from '@utils/common/classNames';

import childFallbackImg from '@assets/images/childFallback.svg';
import fallbackImg from '@assets/images/FallbackImage.svg';

import styles from './SubscriptionCard.module.scss';

/**
 * @interface
 * Base interface for all subscription card types. It includes shared
 * properties such as image data, textual content, and optional behavior
 */
interface CommonSubscriptionCardProps {
  /**
   * A discriminator to identify the card variant. Can be 'childSponsorship', 'addOn', or 'sponsorChild'
   * @property
   */
  cardType: 'childSponsorship' | 'addOn' | 'sponsorChild';
  /**
   * URL of the main image to display
   * @property
   */
  src: string;
  /**
   * Alternate text for the image
   * @property
   */
  alt: string;
  /**
   * Optional fallback image URL in case the main image fails to load
   * @property
   */
  fallbackImage?: string;
  /**
   * Main heading or label for the card
   * @property
   */
  title: string;
  /**
   * Supporting text for the card
   * @property
   */
  description: string;
  /**
   * Optional tab index to control keyboard navigation order
   * @property
   */
  tabIndex?: number;
  /**
   * Optional CSS class name for styling the outer container
   * @property
   */
  containerClassName?: string;
  /**
   * Optional callback function triggered when the card is clicked
   * @property
   */
  onCardClick?: () => void;
}

/**
 * @interface
 * Extends the base card with additional information for child sponsorship,
 * such as location and date of birth.
 */
interface ChildSponsorshipProps extends CommonSubscriptionCardProps {
  /**
   * Specifies that this card is a 'childSponsorship'
   * @property
   */
  cardType: 'childSponsorship';
  /**
   * Optional geographic location of the child
   * @property
   */
  location?: string;
  /**
   * Optional icon name representing the location
   * @property
   */
  locationIconName?: string;
  /**
   * Optional birth date of the child
   * @property
   */
  dateOfBirth?: string;
  /**
   * Optional icon name representing the date of birth
   * @property
   */
  dateOfBirthIconName?: string;
}

/**
 * @interface
 * Represents an add-on item card (non-child sponsorship). Inherits from the
 * base card but explicitly excludes child-related properties
 */
interface AddonCardProps extends CommonSubscriptionCardProps {
  /**
   * Specifies that this card is a 'addOn'
   * @property
   */
  cardType: 'addOn';
  /**
   * Specified as never to disallow these properties for this card type
   * @property
   */
  location?: never;
  /**
   * Specified as never to disallow these properties for this card type
   * @property
   */
  locationIconName?: never;
  /**
   * Specified as never to disallow these properties for this card type
   * @property
   */
  dateOfBirth?: never;
  /**
   * Specified as never to disallow these properties for this card type
   * @property
   */
  dateOfBirthIconName?: never;
}

/**
 * @interface
 * Specialized card for a "sponsor this child" call-to-action. It overrides
 * image and click handler behavior and excludes child detail fields.
 */
interface SponsorChildProps
  extends Omit<
    CommonSubscriptionCardProps,
    'onCardClick' | 'src' | 'fallbackImage' | 'alt'
  > {
  /**
   * Specifies that this card is a 'sponsorChild'
   * @property
   */
  cardType: 'sponsorChild';
  /**
   * Specified as never to disallow these properties for this card type
   * @property
   */
  location?: never;
  /**
   * Specified as never to disallow these properties for this card type
   * @property
   */
  locationIconName?: never;
  /**
   * Specified as never to disallow these properties for this card type
   * @property
   */
  dateOfBirth?: never;
  /**
   * Specified as never to disallow these properties for this card type
   * @property
   */
  dateOfBirthIconName?: never;
  /**
   * Optional (overrides required base props to be optional)
   * @property
   */
  src?: string;
  /**
   * Optional (overrides required base props to be optional)
   * @property
   */
  fallbackImage?: string;
  /**
   * Optional (overrides required base props to be optional)
   * @property
   */
  alt?: string;
  /**
   * Required click handler (mandatory override of optional base prop)
   * @property
   */
  onCardClick: () => void;
}

/**
 * @interface
 * A union type representing all valid subscription card configurations. This enables discriminated
 * unions based on the cardType property to enforce type safety across all card types.
 */
export type SubscriptionCardProps =
  | ChildSponsorshipProps
  | AddonCardProps
  | SponsorChildProps;

/**
 * The SubscriptionCard component is a versatile UI element designed to display
 * different types of subscription-related content cards in a consistent layout.
 * It supports multiple variants including child sponsorship, add-on products,
 * and sponsor-child call-to-actions, using a discriminated union to enforce type-safe props.
 *
 * @group Molecules
 * @category Component
 *
 * @param {SubscriptionCardProps} props - The Subscription Card props
 * @returns {ReactElement} - Subscription Card component
 *
 */
export const SubscriptionCard = ({
  src,
  alt,
  title,
  description,
  location,
  dateOfBirth,
  cardType,
  tabIndex,
  onCardClick,
  fallbackImage = fallbackImg,
  locationIconName = 'icon-location',
  dateOfBirthIconName = 'icon-gift',
  containerClassName,
}: SubscriptionCardProps): ReactElement => {
  const isChildSponsorship = cardType === 'childSponsorship';
  const isSponsorChildType = cardType === 'sponsorChild';
  const imageSource = isSponsorChildType ? childFallbackImg : src;
  const fallbackDescripText = isChildSponsorship
    ? 'Sponsored since &lt;Unknown&gt;'
    : '&lt;Details not available&gt;';
  const name = title || '&lt;Name unavailable&gt;';
  const fallbackAltText = isSponsorChildType
    ? 'Add another sponsorship'
    : 'child avatar unavailable';

  // Function to handle keyboard interactions on the card
  const handleKeyDown = (e: React.KeyboardEvent<HTMLDivElement>) => {
    const { key } = e;
    const keyPressed = key === 'Enter' || key === ' ';
    if (keyPressed && onCardClick) {
      onCardClick();
    }
  };

  return (
    <Card
      className={classNames(
        styles.cardContainer,
        onCardClick ? styles.enabledClick : '',
        containerClassName
      )}
      onClick={onCardClick}
      onKeyDown={handleKeyDown}
      tabIndex={tabIndex || 0}
    >
      <Image
        src={imageSource}
        alt={alt || fallbackAltText}
        aspectRatio="3:2"
        className={styles.imageBlock}
        fallbackSrc={fallbackImage}
      />
      <Box className={styles.profileBlock}>
        <Box className={styles.head}>
          <Heading className={styles.childName}>{name}</Heading>
          <Text className={styles.since}>
            {description || fallbackDescripText}
          </Text>
        </Box>
        {isChildSponsorship && (
          <Box className={styles.foot}>
            {location && (
              <Box className={styles.location}>
                <Icon name={locationIconName} size="16" />
                <Text className={styles.content}>{location}</Text>
              </Box>
            )}
            {dateOfBirth && (
              <Box className={styles.birthDate}>
                <Icon name={dateOfBirthIconName} size="16" />
                <Text className={styles.content}>{dateOfBirth}</Text>
              </Box>
            )}
          </Box>
        )}
      </Box>
    </Card>
  );
};

export default SubscriptionCard;
