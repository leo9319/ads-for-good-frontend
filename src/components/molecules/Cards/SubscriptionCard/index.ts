export type { SubscriptionCardProps } from './SubscriptionCard';
export { default } from './SubscriptionCard';
export { SubscriptionCard } from './SubscriptionCard';
