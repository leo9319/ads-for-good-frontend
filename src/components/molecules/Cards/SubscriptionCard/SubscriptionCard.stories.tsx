import React from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import SubscriptionCard from './SubscriptionCard';
import Theme from '@radix-styles/atoms/Theme';
import childImg from '@assets/images/childImage.svg';
import fallbackImage from '@assets/images/FallbackImage.svg';

// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-export
const meta = {
  title: 'Library/Molecules/Cards/SubscriptionCard',
  component: SubscriptionCard,
  decorators: [
    Story => (
      <Theme>
        <Story />
      </Theme>
    ),
  ],
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/configure/story-layout
    layout: 'centered',
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/writing-docs/autodocs
  tags: ['autodocs'],
  // More on argTypes: https://storybook.js.org/docs/api/argtypes
  argTypes: {
    src: {
      control: 'text',
    },
    alt: {
      control: 'text',
    },
    fallbackImage: {
      control: 'text',
    },
    title: {
      control: 'text',
    },
    description: {
      control: 'text',
    },
    location: {
      control: 'text',
    },
    dateOfBirth: {
      control: 'text',
    },
    cardType: {
      control: 'select',
      options: ['childSponsorship', 'addOn', 'sponsorChild'],
    },
    onCardClick: {
      control: 'object',
    },
    locationIconName: {
      control: 'text',
    },
    dateOfBirthIconName: {
      control: 'text',
    },
  },
  // more on args: https://storybook.js.org/docs/writing-stories/args#args-object
  args: {
    cardType: 'childSponsorship',
  },
} satisfies Meta<typeof SubscriptionCard>;

export default meta;

type Story = StoryObj<typeof SubscriptionCard>;

// More on writing stories with args: https://storybook.js.org/docs/writing-stories/args

export const ChildSponsorship: Story = {
  args: {
    cardType: 'childSponsorship',
    src: childImg,
    alt: 'Child avatar unavailable',
    fallbackImage: fallbackImage,
    title: 'Bryner James',
    description: 'Sponsored since 2025',
    location: 'Guatemala',
    dateOfBirth: 'Oct 15, 2016',
  },
};

export const AddOn: Story = {
  args: {
    cardType: 'addOn',
    src: childImg,
    alt: 'Child avatar unavailable',
    fallbackImage: fallbackImage,
    title: 'Raw Hope',
    description: 'Supporter since 2020',
  },
};

export const SponsorChild: Story = {
  args: {
    cardType: 'sponsorChild',
    title: 'Sponsor a child',
    description: 'Change one more lifeâ€”for good.',
    onCardClick: () => console.log('Clicked'),
  },
};
