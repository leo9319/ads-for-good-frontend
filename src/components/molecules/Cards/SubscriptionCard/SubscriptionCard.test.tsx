import React from 'react';
import { render } from '@testing-library/react';
import SubscriptionCard from './SubscriptionCard';

describe('Subscription Card', () => {
  it('Should Child Sponsorship Variant render successfully', () => {
    const { baseElement } = render(
      <SubscriptionCard
        alt="Child avatar unavailable"
        cardType="childSponsorship"
        dateOfBirth="Oct 15, 2016"
        description="Sponsored since 2025"
        location="Guatemala"
        src="/src/assets/images/childImage.svg"
        title="Bryner James"
      />
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('Should Sponsor a child Variant render successfully', () => {
    const { baseElement } = render(
      <SubscriptionCard
        cardType="sponsorChild"
        title="Bryner James"
        description="Sponsored since 2025"
        onCardClick={() => console.log('clicked')}
      />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
