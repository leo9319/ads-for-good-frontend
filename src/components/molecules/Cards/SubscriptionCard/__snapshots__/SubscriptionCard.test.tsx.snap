// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscription Card Should Child Sponsorship Variant render successfully 1`] = `
<body>
  <div>
    <div
      class="rt-reset rt-BaseCard rt-Card rt-r-size-1 rt-variant-surface cardContainer"
      tabindex="0"
    >
      <div
        data-radix-aspect-ratio-wrapper=""
        style="position: relative; width: 100%; padding-bottom: 66.66666666666667%;"
      >
        <div
          style="position: absolute; top: 0px; right: 0px; bottom: 0px; left: 0px;"
        >
          <img
            alt="Child avatar unavailable"
            class="imgWrapper imageBlock"
            src="/src/assets/images/childImage.svg"
            style="--imgWidth: 100%; --imgHeight: 100%;"
          />
        </div>
      </div>
      <div
        class="rt-Box profileBlock"
      >
        <div
          class="rt-Box head"
        >
          <h1
            class="rt-Heading rt-r-size-6 headingContainer mocked-class-name typography-font-family-default childName"
            id=""
          >
            Bryner James
          </h1>
          <span
            class="rt-Text mocked-class-name typography-font-family-default since"
            id=""
          >
            Sponsored since 2025
          </span>
        </div>
        <div
          class="rt-Box foot"
        >
          <div
            class="rt-Box location"
          >
            <span
              style="border-radius: 0; display: grid; align-items: center; justify-content: center;"
            >
              <span
                class="icon icon-location"
                style="width: 16px; height: 16px;"
              />
            </span>
            <span
              class="rt-Text mocked-class-name typography-font-family-default"
              id=""
            >
              Guatemala
            </span>
          </div>
          <div
            class="rt-Box birthDate"
          >
            <span
              style="border-radius: 0; display: grid; align-items: center; justify-content: center;"
            >
              <span
                class="icon icon-gift"
                style="width: 16px; height: 16px;"
              />
            </span>
            <span
              class="rt-Text mocked-class-name typography-font-family-default"
              id=""
            >
              Oct 15, 2016
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Subscription Card Should Sponsor a child Variant render successfully 1`] = `
<body>
  <div>
    <div
      class="rt-reset rt-BaseCard rt-Card rt-r-size-1 rt-variant-surface cardContainer enabledClick"
      tabindex="0"
    >
      <div
        data-radix-aspect-ratio-wrapper=""
        style="position: relative; width: 100%; padding-bottom: 66.66666666666667%;"
      >
        <div
          style="position: absolute; top: 0px; right: 0px; bottom: 0px; left: 0px;"
        >
          <img
            alt="Add another sponsorship"
            class="imgWrapper imageBlock"
            src="SvgrURL"
            style="--imgWidth: 100%; --imgHeight: 100%;"
          />
        </div>
      </div>
      <div
        class="rt-Box profileBlock"
      >
        <div
          class="rt-Box head"
        >
          <h1
            class="rt-Heading rt-r-size-6 headingContainer mocked-class-name typography-font-family-default childName"
            id=""
          >
            Bryner James
          </h1>
          <span
            class="rt-Text mocked-class-name typography-font-family-default since"
            id=""
          >
            Sponsored since 2025
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;
